### PR TITLE
3352 side nav implement search functionality of links

### LINKED ISSUE #3352: [SideNav] Implement search functionality of links
### What package is this for?

- [x] React
- [ ] Angular

### Summary

We have a feature request from a product team to add search to our `SideNav` component and surfaced in our`SuiteHeader`. This is a legacy product so will have deeper nesting and a hard requirement for searching its navigation links. This should align with our other search capabilities in other components. 

**Additional context** Add any other context or screenshots about the feature
request here.

image

### PR DESCRIPTION
**Pull request by bjornalm**
Closes #3352

**Summary**
Adds search functionality and the possibility to pin links to the SideNav

**Change List (commits, features, bugs, etc)**
- Added search functionality behind prop `hasSearch`. Improves on existing nav search in Maximo by 
  - allowing keyboard navigation on search result
  - show all the parents of a matched leaf, not just the closest one
- Added unit testing for search
- Added possibility to pin items when prop `hasSearch` is true using link prop `isPinned`
- Added unit tests for pinned items
- Updated story with new knobs and data

**Acceptance Test (how to verify the PR)**
The functionality is based on the existing implementation in Maximo with a few differences as mentioned above

**Search**
1. Go to https://deploy-preview-3374--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-sidenav--side-nav-component
2. Enable the knobs "show deep nesting example" and "Enable searching (hasSearch)". 
3. Uncheck knob "Demo pinned link during search".
4. Verify that search matches all leaf node links 
5. Verify that parents of matched links are shown but are not clickable or collapsable 
6. Verify that the string entered in the search input is highlighted in the search result (for the leafs)
7. Verify that text overflow is still properly handled
8. Verify that links are working as before (compare with next)
9. Make sure an empty result list with a msg is displayed when the search returns nothing

**Pinned links**
1. Uncheck knob "Enable searching (hasSearch)" 
2. Check knob "Demo pinned link during search"
3. Verify that a "home" link is shown in the list, but NOT pinned at the top
4. Check the knob "Enable searching (hasSearch)" 
5. Verify that a "home" link is now pinned at the top
6. Test different search scenarios to make sure the pinned home link stays visible without moving

**Regression Test (how to make sure this PR doesn't break old functionality)**
Verify that the stories are still working as before (compare with next) when the new knobs "Enable searching (hasSearch)" and "Demo pinned link during search" are unchecked.

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

**Comment by davidicus** [comment]
There is a z- index issue when scrolling the list under the pinned item and search field. 
image
image
image

**Review by davidicus — APPROVED** [review]

**Comment by davidicus** [comment]
Looks good. Just one issue with the z-index. Great job getting that pumped out.

**Review by kevinsperrine — COMMENTED** [review]
Nice work. just a couple minor issues. Works great in the story!

File: packages/react/src/components/SideNav/SideNav.jsx [hunk 1]
```diff
@@ -120,71 +181,143 @@ const SideNav = ({ links, defaultExpanded, isSideNavExpanded, i18n, testId, ...p
...
+            aria-label={link.metaData.label}
+            onClick={link.metaData.onClick}
+            href={link.metaData.href}
+            renderIcon={link.icon}
+            isActive={link.isActive}
+            data-testid={`${testId}-link-${index}`}
+            className={classnames({
+              [`${iotPrefix}--side-nav__item--is-filtering`]: isFiltering,
+            })}
+            {...link.metaData}
```
[the remark this answers is not in the document]
Comment (bjornalm) [reply packages/react/src/components/SideNav/SideNav.jsx:256]: Very good call

**Comment by bjornalm** [comment]
> There is a z- index issue when scrolling the list under the pinned item and search field. image image
> 
> image

Thanks for testing it, I was using a negative padding which caused this. Felt it in my guts that it was wrong.. should have listened to that :-) I fixed now

**Comment by davidicus** [comment]
@bjornalm was fixed with pinned link but without I still see an issue with the search field and z-index of the links
image

**Pull request merged**